### PR TITLE
Add re-connecting to onClose event

### PR DESCRIPTION
**Pull request by harbu**
StreamrClient was not re-connecting on a clean handshake close (`onClose`) but only on error (`onError`). Extracted re-connect logic into a method and then reused it in `onClose` callback. Added a small integration test to verify that client indeed re-connects and receives messages.

### CODE REVIEW

File: src/main/java/com/streamr/client/StreamrClient.java [hunk 1]
```diff
@@ -126,6 +125,9 @@ public void onMessage(String message) {
                 public void onClose(int code, String reason, boolean remote) {
                     log.info("Connection closed! Code: " + code + ", Reason: " + reason);
                     state = State.Disconnected;
+                    if (!userRequestedDisconnect) {
+                        sleepThenReconnect();
+                    }
                     if (!remote) {
```
Comment (hpihkala) [inline src/main/java/com/streamr/client/StreamrClient.java]: How does the change affect the rest of this method? After `sleepThenReconnect()`, the `StreamrClient.this.onClose()` would still get called, if `!remote`? 

Is there an opportunity to make the previously existing code clearer here? What is `remote` and does it overlap in meaning with `userRequestedDisconnect`? Does `StreamrClient.this.onClose()` get called from anywhere else?

Comment (harbu) [reply src/main/java/com/streamr/client/StreamrClient.java]: Not intentional.

I had trouble finding the JavaDoc but eventually found it via google, according to it: https://www.javadoc.io/doc/org.java-websocket/Java-WebSocket/1.3.8/org/java_websocket/client/WebSocketClient.html
> remote - Returns whether or not the closing of the connection was initiated by the remote host.

I guess I could get rid of the field `boolean userRequestedDisconnect` and assume that `remote == true` is indicative of a user's intention of disconnecting. At least based on method `StreamrClient#onClose` this is the case.

Comment (harbu) [reply src/main/java/com/streamr/client/StreamrClient.java]: Got rid of the boolean field in newest commit

File: src/main/java/com/streamr/client/StreamrClient.java [hunk 2]
```diff
@@ -176,6 +174,16 @@ public WebSocketClient getWebsocket() {
         return websocket;
     }
 
+    private void sleepThenReconnect() {
+        log.warn("Disconnected. Attempting to reconnect in " + options.getReconnectRetryInterval() / 1000 + " seconds.");
+        try {
+            Thread.sleep(options.getReconnectRetryInterval());
+        } catch (InterruptedException e) {
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (harbu) [reply src/main/java/com/streamr/client/StreamrClient.java]: Addressed in newest commit

**Comment by harbu** [comment]
I'm having some problems with threading and re-connection stuff still happening after invoking `disconnect` in test cleanup. Looking into that now.

File: src/main/java/com/streamr/client/StreamrClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kare) [reply src/main/java/com/streamr/client/StreamrClient.java]: @harbu And remove `e.printStackTrace()`?

File: src/main/java/com/streamr/client/StreamrClient.java [hunk 3]
```diff
@@ -245,6 +249,10 @@ public void disconnect() throws ConnectionTimeoutException {
         }
     }
 
+    public void setErrorMessageHandler(ErrorMessageHandler errorMessageHandler) {
+        this.errorMessageHandler = errorMessageHandler;
+    }
+
```
Comment (mthambipillai) [inline src/main/java/com/streamr/client/StreamrClient.java]: For consistency with the way the other parameters are passed, could be passed in `StreamrClientOptions` instead?

**Comment by harbu** [comment]
Decided in peer-to-peer meeting to create a separate issue out of the encountered threading issue.

File: src/main/java/com/streamr/client/StreamrClient.java (hunk 3 above)
Comment (harbu) [reply src/main/java/com/streamr/client/StreamrClient.java]: I agree. However this should probably be done in separate PR.

**Pull request merged**